chain/subscriber.go

```diff
@@ -1,29 +1,496 @@
+		cus = append(cus, v2ContractUpdate(fce, rev, res))
+	})
+	for _, cu := range cus {
+		if err := s.updateContract(tx, cru.State.Index, cu.fcid, cu.prev, cu.curr, cu.resolved, cu.valid); err != nil {
+			return fmt.Errorf("failed to revert v2 contract update: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (s *Subscriber) sync() error {
+	start := time.Now()
+
+	// fetch current chain index
+	index, err := s.cs.ChainIndex(s.shutdownCtx)
+	if err != nil {
+		return fmt.Errorf("failed to get chain index: %w", err)
+	}
+	s.logger.Debugw("sync started", "height", index.Height, "block_id", index.ID)
+
+	// fetch updates until we're caught up
+	var cnt uint64
+	for index != s.cm.Tip() && !s.isClosed() {
+		// fetch updates
+		istart := time.Now()
+		crus, caus, err := s.cm.UpdatesSince(index, updatesBatchSize)
+		if err != nil {
+			return fmt.Errorf("failed to fetch updates: %w", err)
+		}
+		s.logger.Debugw("fetched updates since", "caus", len(caus), "crus", len(crus), "since_height", index.Height, "since_block_id", index.ID, "ms", time.Since(istart).Milliseconds())
+
+		// process updates
+		istart = time.Now()
+		index, err = s.processUpdates(s.shutdownCtx, crus, caus)
+		if err != nil {
+			return fmt.Errorf("failed to process updates: %w", err)
+		}
+		s.logger.Debugw("processed updates successfully", "new_height", index.Height, "new_block_id", index.ID, "ms", time.Since(istart).Milliseconds())
+		cnt++
+	}
+
+	s.logger.Debugw("sync completed", "start_height", index.Height, "block_id", index.ID, "ms", time.Since(start).Milliseconds(), "iterations", cnt)
+	return nil
+}
+
+func (s *Subscriber) processUpdates(ctx context.Context, crus []chain.RevertUpdate, caus []chain.ApplyUpdate) (types.ChainIndex, error) {
+	var index types.ChainIndex
+	if err := s.cs.ProcessChainUpdate(ctx, func(tx ChainUpdateTx) error {
+		// process wallet updates
+		if err := wallet.UpdateChainState(tx, s.walletAddress, caus, crus); err != nil {
+			return fmt.Errorf("failed to process wallet updates: %w", err)
+		}
+
+		// process revert updates
+		for _, cru := range crus {
+			if err := s.revertChainUpdate(tx, cru); err != nil {
+				return fmt.Errorf("failed to revert chain update: %w", err)
+			}
+		}
+
+		// process apply updates
+		for _, cau := range caus {
+			if err := s.applyChainUpdate(tx, cau); err != nil {
+				return fmt.Errorf("failed to apply chain updates: %w", err)
+			}
+		}
+
+		// update chain index
+		index = caus[len(caus)-1].State.Index
+		if err := tx.UpdateChainIndex(index); err != nil {
+			return fmt.Errorf("failed to update chain index: %w", err)
+		}
+
+		// update failed contracts
+		if err := tx.UpdateFailedContracts(index.Height); err != nil {
+			return fmt.Errorf("failed to update failed contracts: %w", err)
+		}
+
+		return nil
+	}); err != nil {
+		return types.ChainIndex{}, fmt.Errorf("failed to process chain update: %w", err)
```

Comment:
This should probably mirror `hostd`'s current `ProcessConsensusChange` and panic. If we can't sync consensus anymore there is no way to recover.